File: src/nucleoid.js
const runtime = require("./runtime");
const context = require("./context");
const process = require("./process");
const _config = require("./config");
const { parseFunction } = require("./lang/estree/parser");
const { generate } = require("./lang/estree/generator");
const chalk = require("chalk");

function start(config = {}) {
  _config.init(config);

  console.log("🌿 " + chalk.greenBright("Nucleoid") + " runtime is started");
  console.log(chalk.blueBright("🌎 Inspired by Nature\n"));

  process.init();

  return new Promise((resolve) => {
    setImmediate(() => {
      context.run();
      resolve();
    });
  });
}

function register(fn) {
  context.load({
    definition: fn.toString(),
    options: { declarative: true },
  });
}

function run(statement, p2, p3) {
  if (typeof statement === "string") {
    const options = p2;
    return runtime.process(`${statement}`, options);
  } else {
    let scope = p2;
    let options = p3;
    const { params, body } = parseFunction(statement.toString());
    scope =
      scope && params.length
        ? `let ${generate(params[0])}=${JSON.stringify(scope)};`
        : "";
    return runtime.process(`${scope}${generate(body)}`, options);
  }
}

module.exports.start = start;
module.exports.register = register;
module.exports.run = run;

File: cpp/src/Nucleoid.cpp
#include "nucleoid/Nucleoid.hpp"
#include "nucleoid/Config.hpp"
#include "nucleoid/Runtime.hpp"
#include <chrono>
#include <ctime>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>

namespace nucleoid {

void start(const Config& cfg) {
    configInit(cfg);
    std::cout << "Nucleoid runtime is started\n";
    std::cout << "Inspired by Nature\n";
}

nlohmann::json run(const std::string& statement, Options options) {
    Config cfg = configGet();
    options.declarative = options.declarative || cfg.options.declarative;

    Data data = runtimeProcess(statement, options);

    if (options.details) {
        nlohmann::json j;
        j["string"]      = data.string;
        j["declarative"] = data.declarative;
        j["value"]       = data.result.value;
        j["time"]        = data.time;
        j["date"]        = data.date;
        j["error"]       = data.error;
        j["events"]      = nlohmann::json::array();
        for (auto& e : data.events) {
            j["events"].push_back({ {"topic", e.topic}, {"data", e.data} });
        }
        return j;
    }

    return data.result.value;
}

void registerFn(const std::string& fnSource) {
    Options opts;
    opts.declarative = true;
    try {
        runtimeProcess(fnSource, opts);
    } catch (...) {}
}

void registerContext(const nlohmann::json& entry) {
    std::string def = entry.value("definition", "");
    if (def.empty()) return;
    Options opts;
    opts.declarative = entry.contains("options") &&
                       entry["options"].value("declarative", false);
    try {
        runtimeProcess(def, opts);
    } catch (...) {}
}

} // namespace nucleoid
